swarm/src/handler/from_fn.rs

```diff
@@ -0,0 +1,349 @@
+    protocol: &'static str,
+    inbound_streams_limit: usize,
+    pending_dial_limit: usize,
+    on_new_inbound: impl Fn(NegotiatedSubstream, &mut TState) -> TInboundFuture + Send + 'static,
+    on_new_outbound: impl Fn(NegotiatedSubstream, &mut TState, TOutboundOpenInfo) -> TOutboundFuture
+        + Send
+        + 'static,
+) -> FromFn<TInbound, TOutbound, TOutboundOpenInfo, TState>
+where
+    TInboundFuture: Future<Output = TInbound> + Send + 'static,
+    TOutboundFuture: Future<Output = TOutbound> + Send + 'static,
+    TState: Default,
+{
+    FromFn {
+        protocol,
+        inbound_streams_limit,
+        pending_dial_limit,
+        inbound_streams: FuturesUnordered::default(),
+        outbound_streams: FuturesUnordered::default(),
+        on_new_inbound: Box::new(move |stream, state| on_new_inbound(stream, state).boxed()),
+        on_new_outbound: Box::new(move |stream, state, info| {
+            on_new_outbound(stream, state, info).boxed()
+        }),
+        idle_waker: None,
+        pending_dials: VecDeque::default(),
+        failed_open: VecDeque::default(),
+        state: TState::default(),
+    }
+}
+
+#[derive(Debug)]
+pub enum OutEvent<I, O, OpenInfo> {
+    InboundFinished(I),
+    OutboundFinished(O),
+    FailedToOpen(OpenError<OpenInfo>),
+}
+
+#[derive(Debug)]
+pub enum OpenError<OpenInfo> {
+    Timeout(OpenInfo),
+    LimitExceeded(OpenInfo),
+    NegotiationFailed(OpenInfo, NegotiationError),
+}
+
+impl<OpenInfo> fmt::Display for OpenError<OpenInfo> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            OpenError::Timeout(_) => write!(f, "opening new substream timed out"),
+            OpenError::LimitExceeded(_) => write!(f, "limit for pending dials exceeded"),
+            OpenError::NegotiationFailed(_, _) => Ok(()), // Don't print anything to avoid double printing of error.
+        }
+    }
+}
+
+impl<OpenInfo> Error for OpenError<OpenInfo>
+where
+    OpenInfo: fmt::Debug,
+{
+    fn source(&self) -> Option<&(dyn Error + 'static)> {
+        match self {
+            OpenError::Timeout(_) => None,
+            OpenError::LimitExceeded(_) => None,
+            OpenError::NegotiationFailed(_, source) => Some(source),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum InEvent<TState, TOutboundOpenInfo> {
+    UpdateState(TState),
+    NewOutbound(TOutboundOpenInfo),
+}
+
+pub struct FromFn<TInbound, TOutbound, TOutboundInfo, TState> {
+    protocol: &'static str,
+
+    inbound_streams: FuturesUnordered<BoxFuture<'static, TInbound>>,
+    outbound_streams: FuturesUnordered<BoxFuture<'static, TOutbound>>,
+
+    on_new_inbound:
+        Box<dyn Fn(NegotiatedSubstream, &mut TState) -> BoxFuture<'static, TInbound> + Send>,
+    on_new_outbound: Box<
+        dyn Fn(NegotiatedSubstream, &mut TState, TOutboundInfo) -> BoxFuture<'static, TOutbound>
+            + Send,
+    >,
+
+    idle_waker: Option<Waker>,
```

Comment:
Currently implementations of `ConnectionHandler` depend on being `poll`ed after any of the `inject_xxx` methods are called. I don't think that implicit assumption is great. I prefer the approach you take here, making it explicit via `idle_waker`. The only downside is the additional complexity.

Reply:
Right, I didn't even think about it but of course, with our `loop`s, we implicitly poll components again after we have called functions on them.

`ConnectionHandler` is not going to be used outside of `NetworkBehaviour` so I think it is fine to depend on this and remove the complexity from here.